File: tcg_inventory_api/src/main/java/com/jordansimsmith/tcginventory/Games.java
package com.jordansimsmith.tcginventory;

import java.util.Map;
import java.util.Set;

public class Games {
  public record Game(
      String id,
      String externalSource,
      String fetchTcgExternalReferenceField,
      Set<String> finishes) {
    public Game {
      finishes = Set.copyOf(finishes);
    }
  }

  public static final Game MAGIC_THE_GATHERING =
      new Game("mtg", "scryfall", "scryfallId", Set.of("normal", "foil", "etched"));

  private static final Map<String, Game> GAMES =
      Map.of(MAGIC_THE_GATHERING.id(), MAGIC_THE_GATHERING);

  public static Game get(String id) {
    var game = GAMES.get(id);
    if (game == null) {
      throw new IllegalArgumentException("unsupported game: " + id);
    }
    return game;
  }
}

File: tcg_inventory_web/src/domain/games.ts
export const MAGIC_THE_GATHERING = {
  id: 'mtg',
  label: 'Magic: The Gathering',
  externalSource: 'scryfall',
  finishes: ['normal', 'foil', 'etched'],
} as const;

export const GAMES = [MAGIC_THE_GATHERING] as const;

export type GameId = (typeof GAMES)[number]['id'];
export type Finish = (typeof GAMES)[number]['finishes'][number];

export function getGame(gameId: string) {
  const game = GAMES.find(({ id }) => id === gameId);
  if (!game) {
    throw new Error(`Unsupported game: ${gameId}`);
  }
  return game;
}

export function gameLabel(gameId: string): string {
  return getGame(gameId).label;
}
